mark phase in forked child

FraMecca opened the pull request:
The aim of this pr is to match the forking capabilities of the sociomantic GC with the heap growth strategy of the current Druntime GC.
It should be noted that the code allocates new memory pool to be able to fulfil the memory request instantly to avoid blocking the thread that triggered the collection in case of memory exhaustion.

Commit:
mark phase in forked child

llucax commented on src/gc/impl/conservative/gc.d:
```diff
@@ -2393,7 +2407,36 @@ struct Gcx
+                                                  cast(int) block);
+                    markProcPid = 0;
+                    // process GC marks then sweep
+                    thread_suspendAll();
```
Why are you suspending all threads here?

FraMecca replied:
https://github.com/dlang/druntime/blob/17cc32231dfc800a54d45745f2b36589a8a98ae3/src/core/thread.d#L3187

It seems to me that processGCMarks must be called after pausing every thread

llucax replied:
Oh, OK, this is new in D2.

llucax commented on src/gc/impl/conservative/gc.d:
```diff
@@ -2393,7 +2407,36 @@ struct Gcx
+                case WRes.ERROR:
+                    debug(COLLECT_PRINTF) printf("\t\tmark proc ERROR\n");
+                    markProcPid = 0;
+                    return fullcollect(nostack, true); // Try to keep going without forking
```
Are you planning to completely disable the forking behavior in the future if something fails or do you plan to handle these kind of (nasty) errors differently compared to my implementation?

But in any case, I find this a bit risky, you are calling `fullcollect()` recursivelly, why doing so when you can just continue with this call?

llucax commented on src/gc/impl/conservative/gc.d:
```diff
@@ -2393,7 +2407,36 @@ struct Gcx
         debug(COLLECT_PRINTF) printf("Gcx.fullcollect()\n");
         //printf("\tpool address range = %p .. %p\n", minAddr, maxAddr);
-
+        if (collectInProgress)
```
I would like to see more documentation here, see the [comments in my implementation](https://github.com/sociomantic-tsunami/tangort/blob/v1.x.x/src/gc/cdgc/gc.d#L888-L894), what's going on might not be trivial for a casual reader, so better to be as clear as possible on what's going on.

llucax commented on src/gc/impl/conservative/gc.d:
```diff
@@ -2414,12 +2457,39 @@ struct Gcx
             }
 
-            markAll(nostack);
+            if (block)
```
Same about comments in here, see the [comments in my implementation](https://github.com/sociomantic-tsunami/tangort/blob/v1.x.x/src/gc/cdgc/gc.d#L931-L939). And this goes in general, at least for my taste this code is a bit underdocumented. Again, stuff done in here might be hard to follow due to the concurrency, so it's better to have good comments in the code.

llucax commented on src/gc/impl/conservative/gc.d:
```diff
@@ -2414,12 +2457,39 @@ struct Gcx
+                {
+                    case -1: // fork() failed, retry without forking
+                        markProcPid = 0;
+                        return fullcollect(nostack, true);
```
Same about the recursive call instead of just continuing with this call. Is this really safe? Maybe explain why in a comment?

llucax commented on src/gc/impl/conservative/gc.d:
```diff
@@ -2414,12 +2457,39 @@ struct Gcx
+                            markAll(nostack);
+                            _Exit(0);
+                            break; // bogus
+                    default: // the parent
```
The `block` variable is not properly handled here. Is this intentional? See https://github.com/sociomantic-tsunami/tangort/blob/v1.x.x/src/gc/cdgc/gc.d#L952-L970

FraMecca replied:
That block of code is not entered if the block variable is set to true (line 2461) and the mark is done in the current process.
Anyway, I updated the code to take into consideration a shouldFork boolean

llucax reviewed (commented):
A few comments, but in general, I also wonder if this was actually tested and verified it is working or if it's just a draft implementation.

It might be useful to mention the maturity in the PR.

Commit:
Added option for disabling fork()

In the future the boolean variable will be moved to runtime options
In the future the boolean variable will be moved to runtime options

Commit:
use goto instead of calling the function recursively

Commit:
Added comments

Commit:
disableFork is nothrow

FraMecca commented:
I added some small commits to reflect your changes.
Regarding the maturity:
tests are all successful. Valgrind shows the same frees as the current runtime.

However, I'd like to point out that when the druntime does a last collection before exiting and that should run blocking or without forking. I'd prefer to discuss that in a future PR.

llucax commented on src/gc/impl/conservative/gc.d:
```diff
@@ -2477,16 +2480,30 @@ struct Gcx
-                        markProcPid = 0;
-                        return fullcollect(nostack, true);
+                        disableFork();
+                        goto MARK;
```
OK, it seems like the logic in this implementation is a bit more complicated than it was for D1 where we could just go on without `goto`s. I have the feeling a `goto` like this might not pass upstream review, so feel free to handle this as you think it's best.

FraMecca replied:
I actually used goto because they are used a lot in the druntime code.

llucax commented on src/gc/impl/conservative/gc.d:
```diff
@@ -2467,10 +2473,23 @@ MARK:
                 start = stop;
             }
 
-            if (block)
```
This change seems to be misplaced in this commit, which should only be about adding comments.

FraMecca replied:
Git forced me to stage the comment and the if condition (on the comment below) as replacement for just this line.
The error was mine that I wrote the comments and changed the line thinking that I could commit it line by line.
I should have changed the if condition, pushed, then added the comments.

llucax replied:
You can always edit commits as much as you want/need. Just use `git rebase -i`, use `edit` on the commit you want to edit (for example the one where code is in the commit adding documentation to remove the changed code), then go to the commit where the code should be and add it back (if you are not forced to do so due to a conflict).

In any case, if you just prefer to squash everything together to avoid spending too much time on this I think it's OK too.

llucax commented on src/gc/impl/conservative/gc.d:
```diff
@@ -2467,10 +2473,23 @@ MARK:
+            // Forking is not enabled.
+            // This is a standard stop the world collection
+            // and the mark is done in this thread.
+            if (!shouldFork)
```
Same, changed in the wrong commit.

Commit:
changed MARK label to Lmark

FraMecca commented:
I changed the mark label to better reflect the style that is used in the rest of the code.
Should I squash the commits into one and merge?
Or should I go on and modify/test something else?

llucax reviewed (approved):
LGTM. Again, I would love to see some number for benchmarks on the commit message and/or the PR description but I thinks the basics are done.
